common/crypto/dkg/src/dealing.rs

```diff
@@ -0,0 +1,485 @@
+    // allowing the dealer to skip decryption of its own share if it was also one of the receivers
+    pub fn create(
+        mut rng: impl RngCore,
+        params: &Params,
+        dealer_index: NodeIndex,
+        threshold: Threshold,
+        epoch: Epoch,
+        // BTreeMap ensures the keys are sorted by their indices
+        receivers: &BTreeMap<NodeIndex, PublicKey>,
+    ) -> (Self, Option<Share>) {
+        assert!(threshold > 0);
+
+        let polynomial = Polynomial::new_random(&mut rng, threshold - 1);
+        let mut shares = receivers
+            .keys()
+            .map(|&node_index| polynomial.evaluate_at(&Scalar::from(node_index)).into())
+            .collect::<Vec<_>>();
+
+        let remote_share_key_pairs = shares
+            .iter()
+            .zip(receivers.values())
+            .map(|(share, key)| (share, key))
+            .collect::<Vec<_>>();
+        let ordered_public_keys = receivers.values().copied().collect::<Vec<_>>();
+
+        let (ciphertexts, hazmat) =
+            encrypt_shares(&remote_share_key_pairs, epoch, params, &mut rng);
+
+        // create proofs of knowledge
+        let chunking_instance = proof_chunking::Instance::new(&ordered_public_keys, &ciphertexts);
+        let proof_of_chunking =
+            ProofOfChunking::construct(&mut rng, chunking_instance, hazmat.r(), &shares)
+                .expect("failed to construct proof of chunking");
+
+        let combined_ciphertexts = ciphertexts.combine_ciphertexts();
+        let mut combined_r = hazmat.combine_rs();
+        let combined_rr = ciphertexts.combine_rs();
+
+        let public_coefficients = polynomial.public_coefficients();
+        let sharing_instance = proof_sharing::Instance::new(
+            receivers,
+            &public_coefficients,
+            &combined_rr,
+            &combined_ciphertexts,
+        );
+        let proof_of_sharing =
+            ProofOfSecretSharing::construct(&mut rng, sharing_instance, &combined_r, &shares)
+                .expect("failed to construct proof of secret sharing");
+
+        combined_r.zeroize();
+
+        let dealing = Dealing {
+            public_coefficients,
+            ciphertexts,
+            proof_of_chunking,
+            proof_of_sharing,
+        };
+
+        let dealers_key_index = receivers
+            .keys()
+            .position(|node_index| node_index == &dealer_index);
+        if let Some(dealer_key_index) = dealers_key_index {
+            let dealers_share = shares.remove(dealer_key_index);
+            shares.zeroize();
+            (dealing, Some(dealers_share))
+        } else {
+            (dealing, None)
+        }
+    }
+}
+
+impl Dealing {
+    // rather than returning a bool for whether the dealing is valid or not, a Result is returned
+    // instead so that we would have more information regarding a possible failure cause
+    pub fn verify(
```

Comment:
Here we also add as an optional input argument the secret value known by the initiating validator (so we can verify a resharing if it happened)

Reply:
Correction: "optional input argument the secret value" -> "optional input argument the partial verification key"